src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,492 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues
+{
+    internal abstract partial class AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private sealed partial class SymbolStartAnalyzer
+        {
+            private sealed partial class BlockAnalyzer
+            {
+                private readonly SymbolStartAnalyzer _symbolStartAnalyzer;
+                private readonly Options _options;
+
+                /// <summary>
+                /// Indicates if the operation block has an <see cref="IDelegateCreationOperation"/> or an <see cref="IAnonymousFunctionOperation"/>.
+                /// We use this value in <see cref="ShouldAnalyze(IOperation, ISymbol)"/> to determine whether to bail from analysis or not.
+                /// </summary>
+                private bool _hasDelegateCreationOrAnonymousFunction;
+
+                /// <summary>
+                /// Indicates if a delegate instance escaped this operation block, via an assignment to a field or a property symbol.
+                /// that can be accessed outside this executable code block.
+                /// We use this value in <see cref="ShouldAnalyze(IOperation, ISymbol)"/> to determine whether to bail from analysis or not.
+                /// </summary>
+                private bool _delegateAssignedToFieldOrProperty;
+
+                /// <summary>
+                /// Indicates if the operation block has an <see cref="IConversionOperation"/> with a delegate type as it's source type
+                /// and a non-delegate type as it's target.
+                /// We use this value in <see cref="ShouldAnalyze(IOperation, ISymbol)"/> to determine whether to bail from analysis or not.
+                /// </summary>
+                private bool _hasConversionFromDelegateTypeToNonDelegateType;
+
+                private BlockAnalyzer(SymbolStartAnalyzer symbolStartAnalyzer, Options options)
+                {
+                    _symbolStartAnalyzer = symbolStartAnalyzer;
+                    _options = options;
+                }
+
+                public static void Analyze(OperationBlockStartAnalysisContext context, SymbolStartAnalyzer symbolStartAnalyzer)
+                {
+                    if (HasSyntaxErrors() || context.OperationBlocks.IsEmpty)
```

Comment:
when do you get 'empty' operation blocks?

Reply:
standard disclaimed about passing data implicitly over long distances being hard to reason about :)